### ISSUE TITLE
Reconnect after a disconnect does not seem to work.

### ISSUE THREAD

**Issue by pallavg** [jasonrbriggs/stomp.py#311]
Good afternoon,

This issue might be similar to #308 and #294. Please advise. I followed the documentation/tutorial on how to setup a listener to deal with disconnects. I am using stomp.py 6.0.0 (latest), python 3.7.6, and django 3.0.5. For the message broker, I am using ActiveMQ.

My message broker class looks like this:

```
# Message broker (ActiveMQ/STOMP) service to send/retrieve messages.
#
# --------------------------------------------- MODULE IMPORT ---------------------------------------------------------#
import atexit
import json
import logging
import os
import signal
import ssl
import sys
from pathlib import Path

import stomp


# --------------------------------------------- DEVELOPMENT CODE ------------------------------------------------------#
log = logging.getLogger(__name__)


# Cleanup function.
def _cleanup(signum, frame):
    sys.exit(0)

# On Windows, need to handle SIGBREAK as well.
signal.signal(signal.SIGINT, _cleanup)
if sys.platform.startswith('win'):
    signal.signal(signal.SIGBREAK, _cleanup)  # pylint: disable=no-member


def get_credentials():
    """Read ActiveMQ credentials.

    Returns:
        Dictionary
    """

    # Credentials. Stomp expects list of tuples for host/ports.
    secret_key_fp = Path(__file__).parent.parent.joinpath('secret_data.json')
    with secret_key_fp.open() as f_in:
        secret_data = json.load(f_in)

    for k in secret_data.keys():
        if k.startswith('django_activemq'):
            secret_data[k]['hosts'] = [tuple(x) for x in secret_data[k]['hosts']]

    return secret_data


def _conn_cleanup(c):
    """Close connection."""

    if c is not None and c.is_connected():
        log.info('Close ActiveMQ connection')

        # Remove connection failure listener to prevent reconnect during disconnect!
        c.remove_listener('connection_failure_listener')
        c.disconnect()


class ConnectionFailureListener(stomp.ConnectionListener):
    """This class is used to re-establish a connection to the ActiveMQ server in case of disconnect."""

    def __init__(self, conn):
        """Initialization.

        Args:
            conn -- Connection object
        """

        self._conn = conn

    def on_disconnected(self):
        """Reconnect to MQ if there is a connection failure."""

        log.info('Attempt reconnect to ActiveMQ')
        self._conn.connect()


    def on_error(self, headers, body):

        log.info('got error')
        import pprint
        pprint.pprint(headers)
        print()
        pprint.pprint(body)

    def on_message(self, headers, body):
        log.info('got msg')
        import pprint
        pprint.pprint(headers)
        print()
        pprint.pprint(body)

class MsgBkr(object):

    def __init__(self, producer=True, conn_name='django_activemq'):
        """Initialization.

        Args:
            producer -- Whether this connection is a producer (writer) or consumer (reader)
        """

        # 'django_activemq' is the global producer queue. Consumers have their own queue which user should provide.
        if producer:
            assert conn_name == 'django_activemq'
        else:
            assert conn_name != 'django_activemq'

        self._producer = producer
        self._data = get_credentials()
        self._conn_name = conn_name
        conn = self._create_connection()
        self._conn = conn

        # Add listeners.
        self._conn.set_listener('connection_failure_listener', ConnectionFailureListener(self._conn))

        # To prevent stale TCP connection, make sure we disconnect when we exit or get a SIGINT/SIGBREAK.
        atexit.register(_conn_cleanup, conn)

    def _create_connection(self):
        """Create connection instance to ActiveMQ message broker.

        Returns:
            A Stomp connection object or None
        """

        # Type of connection.
        if self._producer:
            conn_type, conn_purpose = 'producer', 'sending'
        else:
            conn_type, conn_purpose = 'consumer', 'receiving'

        # Create SSL connection.
        log.info('Create ActiveMQ (%s) connection for (%s) messages', conn_type, conn_purpose)
        try:
            conn = stomp.Connection(self._data[self._conn_name]['hosts'])
            conn.set_ssl(for_hosts=self._data[self._conn_name]['hosts'], ssl_version=ssl.PROTOCOL_TLS)
        except stomp.exception.ConnectFailedException:
            log.critical('Unable to create ActiveMQ connection; check host ip/port and/or user/pass')
            conn = None

        return conn

    def connect(self):
        """Connect to ActiveMQ with user credentials.

        Note: This function should be called AFTER all listeners have been registered (w/ set_listener).
        """

        if self._conn is None:
            log.critical('ActiveMQ connection is None')
            return

        # If already connected, disconnect first. Then, reconnect.
        if self._conn.is_connected():
            print('going to disconnect')
            self._conn.disconnect()

        # Producers can write to queue. Consumers can only read from queue.
        # NOTE: In both cases, we will use the '_rw' user. If there is a long period of inactivity, the queue gets
        # removed in DBaaS (not deleted). If we try to access with '_ro' user, we get an error saying we do not
        # have permission to create queue with '_ro' user.
        k = 'rw'
        username = self._data[self._conn_name]['user_' + k]
        password = self._data[self._conn_name]['pass_' + k]

        # Connect to message broker with user/pass.
        print('username', username)
        print('password', password)
        self._conn.connect(username, password, wait=True)

        print('going to subscribe)')
        # Consumers (clients) subscribe to listen for messages and must send acknowledgment for EVERY message.
        if not self._producer:
            self._conn.subscribe(destination=self._data[self._conn_name]['queue'], id=1, ack='client-individual')

    def is_connected(self):
        """Determine if connected.

         Returns:
             Boolean
        """

        return self._conn is not None and self._conn.is_connected()

    def close(self):
        """Close if connected."""

        _conn_cleanup(self._conn)

    def put(self, msg, **kwargs):
        """Put message on ActiveMQ.

        Args:
            msg    -- Message (serialized JSON)
            kwargs -- Additional message headers

        Note:
            This action is legal for producers only.
        """

        if not self._producer:
            log.error('ActiveMQ object is a consumer; does not have permission to send/write messages')
            return

        if self._conn is None:
            log.critical('ActiveMQ connection is None')
            return

        # Ensure persistent=True so that messages are stored in MQ and not lost due to crash/power failure.
        #
        # NOTE: In development mode, we do not want to use the production MQ for testing. We have a separate MQ
        # (django_activemq_dev) that should be used for retrieving messages. For routing messages to this queue,
        # need to set dev=true in the message header.
        headers = {'persistent': 'true',
                   'dev': str(os.environ.get('FLASH_DJANGO_MODE', 'Dev') == 'Dev').lower()}
        headers.update(kwargs)

        # Write serialized JSON message to ActiveMQ queue.
        self._conn.send(destination=self._data[self._conn_name]['queue'],
                        body=msg,
                        content_type='application/json; charset=utf-8',
                        headers=headers)

    def set_listener(self, name, lstnr, **kwargs):
        """Add listener to connection object.

        Args:
            name   -- Listener (unique) name
            lstnr  -- Listener class
            kwargs -- Args to pass to Listener
        """

        if self._conn.get_listener(name) is not None:
            log.error('Attempt to add listener with an already registered name=%s', name)
            return

        self._conn.set_listener(name, lstnr(self._conn, **kwargs))

    def remove_listener(self, name):
        """Remove an existing listener.

        Args:
            name -- Listener name
        """

        self._conn.remove_listener(name)
```

The above class works fine for both producer/consumers. However, sometimes due to idleness/inactivity and/or broker failover (There are 3 brokers to ensure high availability), the connection gets lost. When attempting to reconnect, I am getting a bizarre error from ActiveMQ regarding invalid user/pass which doesn't make sense (otherwise, the connection would not have been made in the first place?). The reconnection used to work in the previous version. Here is a stack dump showing the issue. 

If I kill this process and start a new one, the connection/loop runs fine. Please advise on a resolution / workaround (if any). Thanks.

```
> 
> apps.traceconvert.forms INFO     - Send request to ActiveMQ
> stomp.py INFO     - Sending frame: 'SEND'
> apps.traceconvert.forms INFO     - Send email notification to user admin
> django.server INFO     - "POST /traceconvert/ HTTP/1.1" 302 0
> django.server INFO     - "POST /traceconvert/ HTTP/1.1" 302 0
> django.server INFO     - "GET /traceconvert/ HTTP/1.1" 200 42645
> django.server INFO     - "GET /traceconvert/ HTTP/1.1" 200 42645
> django.server INFO     - "GET /static/css/sb-admin-2.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/css/tvpv_portal.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/css/sb-admin-2.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/css/tvpv_portal.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/js/sb-admin-2.min.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/js/sb-admin-2.min.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/js/tvpv_portal.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/js/tvpv_portal.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/flash_logo.svg HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/flash_logo.svg HTTP/1.1" 304 0
> django.server INFO     - "GET /static/traceconvert/js/traceconvert_form.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/img/ajax-loader.gif HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/img/ajax-loader.gif HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/js/toolbar.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/js/toolbar.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/css/toolbar.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/css/toolbar.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/traceconvert/js/traceconvert_form.js HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/css/print.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/css/print.css HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/img/djdt_vertical.png HTTP/1.1" 304 0
> django.server INFO     - "GET /static/debug_toolbar/img/djdt_vertical.png HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/site.webmanifest HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/site.webmanifest HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/android-chrome-192x192.png HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/android-chrome-192x192.png HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/favicon-16x16.png HTTP/1.1" 304 0
> django.server INFO     - "GET /static/img/favicons/favicon-16x16.png HTTP/1.1" 304 0
> stomp.py INFO     - Receiver loop ended
> services.msgbkr INFO     - Attempt reconnect to ActiveMQ
> stomp.py INFO     - attempt reconnection (True, None, 0)
> stomp.py INFO     - Attempting connection to host  XX.XX.XX.XXX, port 8501
> stomp.py WARNING  - Could not connect to host  XX.XX.XX.XXX, port 8501
> Traceback (most recent call last):
>   File "C:\Anaconda3\lib\site-packages\stomp\transport.py", line 727, in attempt_connection
>     self.socket = socket.create_connection(host_and_port, self.__timeout)
>   File "C:\Anaconda3\lib\socket.py", line 728, in create_connection
>     raise err
>   File "C:\Anaconda3\lib\socket.py", line 716, in create_connection
>     sock.connect(sa)
> ConnectionRefusedError: [WinError 10061] No connection could be made because the target machine actively refused it
> stomp.py INFO     - Attempting connection to host XX.XX.XX.XXX, port 8501
> stomp.py INFO     - Established connection to host  XX.XX.XX.XXX, port 8501
> stomp.py INFO     - Starting receiver loop (<Thread(Thread-17, started daemon 22720)>)
> stomp.py INFO     - Created thread <Thread(Thread-17, started daemon 22720)> using func <function default_create_thread at 0x0000027CBAF7F3A8>
> stomp.py INFO     - Sending frame: 'STOMP'
> stomp.py INFO     - Received frame: 'ERROR', len(body)=2129
> services.msgbkr INFO     - got error
> {'content-type': 'text/plain',
>  'message': 'User name [null] or password is invalid.'}
> 
> ('java.lang.SecurityException: User name [null] or password is invalid.\n'
>  '\tat '
>  'org.apache.activemq.security.SimpleAuthenticationBroker.authenticate(SimpleAuthenticationBroker.java:103)\n'
>  '\tat '
>  'org.apache.activemq.security.SimpleAuthenticationBroker.addConnection(SimpleAuthenticationBroker.java:71)\n'
>  '\tat '
>  'org.apache.activemq.broker.BrokerFilter.addConnection(BrokerFilter.java:98)\n'
>  '\tat '
>  'org.apache.activemq.broker.MutableBrokerFilter.addConnection(MutableBrokerFilter.java:103)\n'
>  '\tat '
>  'org.apache.activemq.broker.TransportConnection.processAddConnection(TransportConnection.java:849)\n'
>  '\tat '
>  'org.apache.activemq.broker.jmx.ManagedTransportConnection.processAddConnection(ManagedTransportConnection.java:77)\n'
>  '\tat '
>  'org.apache.activemq.command.ConnectionInfo.visit(ConnectionInfo.java:139)\n'
>  '\tat '
>  'org.apache.activemq.broker.TransportConnection.service(TransportConnection.java:333)\n'
>  '\tat '
>  'org.apache.activemq.broker.TransportConnection$1.onCommand(TransportConnection.java:197)\n'
>  '\tat '
>  'org.apache.activemq.transport.MutexTransport.onCommand(MutexTransport.java:45)\n'
>  '\tat '
>  'org.apache.activemq.transport.AbstractInactivityMonitor.onCommand(AbstractInactivityMonitor.java:300)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.StompTransportFilter.sendToActiveMQ(StompTransportFilter.java:97)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.ProtocolConverter.sendToActiveMQ(ProtocolConverter.java:202)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.ProtocolConverter.onStompConnect(ProtocolConverter.java:774)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.ProtocolConverter.onStompCommand(ProtocolConverter.java:265)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.StompTransportFilter.onCommand(StompTransportFilter.java:85)\n'
>  '\tat '
>  'org.apache.activemq.transport.TransportSupport.doConsume(TransportSupport.java:83)\n'
>  '\tat '
>  'org.apache.activemq.transport.tcp.SslTransport.doConsume(SslTransport.java:108)\n'
>  '\tat '
>  'org.apache.activemq.transport.stomp.StompSslTransportFactory$1$1.doConsume(StompSslTransportFactory.java:70)\n'
>  '\tat '
>  'org.apache.activemq.transport.tcp.TcpTransport.doRun(TcpTransport.java:233)\n'
>  '\tat '
>  'org.apache.activemq.transport.tcp.TcpTransport.run(TcpTransport.java:215)\n'
>  '\tat java.lang.Thread.run(Thread.java:748)\n')
> stomp.py INFO     - Receiver loop ended
> apps.traceconvert.forms INFO     - Send request to ActiveMQ
> stomp.py INFO     - Sending frame: 'SEND'
> django.server ERROR    - "POST /traceconvert/ HTTP/1.1" 500 197786
> django.server ERROR    - "POST /traceconvert/ HTTP/1.1" 500 197786
```

**Comment by jasonrbriggs** [comment]
@pallavg Can you try with the latest on the dev branch?  There was an issue with disconnect notifications being sent twice, which messed up the reconnection -- there's a fix in dev for this now.

**Comment by jasonrbriggs** [comment]
Hopefully fixed in 6.1.0 - re-open issue if not.
